### ISSUE TITLE
Unable to use contour for typical wildcard TLS setup: `IngressSpec.tls.secretName` seems to be required, proxy protocol support is not working

### ISSUE THREAD

**Issue by artiommocrenco** [projectcontour/contour#4556]
**What steps did you take and what happened:**
Not willing to use contour CRDs.

Using the same `networking.k8s.io/v1` Ingress resources that worked with nginx-ingress. Expected them to work with contour.

I am testing out bitnami/contour chart with the following helm values:
```yaml
contour:
  extraArgs:
    - --use-proxy-protocol
configInline:
  network:
    num-trusted-hops: 1
  tls:
    fallback-certificate:
      name: wildcard-tls
      namespace: contour
```

where `wildcard-tls` secret is created by cert-manager for `*.example.com`

I have Ingress with similar IngressSpec:

```
rules:
- host: app.example.com
  http:
    paths:
    - backend:
        service:
          name: app-ui
          port:
            name: http
      path: /
      pathType: Prefix
tls:
- hosts:
  - app.example.com
```

This produces the following:

1. `unresolved secret reference` contour error

```
time="2022-06-06T12:06:34Z" level=error msg="unresolved secret reference" context=IngressProcessor error="Secret not found" name=app-ui namespace=app secret=app/
```

2. `envoy.http.original_ip_detection.xff` envoy error

```
[2022-06-06 11:51:46.111][1][warning][config] [source/common/config/grpc_subscription_impl.cc:126] gRPC config for type.googleapis.com/envoy.config.listener.v3.Listener rejected: Error adding/updating listener(s) ingress_http: Didn't find a registered implementation for name: 'envoy.http.original_ip_detection.xff'
```

3. Envoy HTTPS endpoints are not starting at all, only HTTP endpoints do (and only if I remove `configInline.network.num-trusted-hops` setting)

**What did you expect to happen:**

1. Not specifying `IngressSpec.tls.secretName` should not produce a log message of type `error`. `IngressSpec.tls.secretName` is optional as per [Kubernetes API Reference](https://kubernetes.io/docs/reference/kubernetes-api/service-resources/ingress-v1/#IngressSpec). 
2. There should not be any `envoy.http.original_ip_detection.xff` envoy-related errors
3. Envoy HTTPS endpoints must be functional and use the configured wildcard certificate as per `fallback-certificate`
4. App must then be available both via `http://app.example.com/` and `https://app.example.com/`

**Environment:**

- Contour version: 
```
contour-7.10.1              	1.20.1     
```
- Kubernetes version: (use `kubectl version`): 1.24

**Comment by sunjayBhatia** [comment]
Unfortunately we never prioritized/implemented fallback integration with Ingress: https://github.com/projectcontour/contour/issues/2453

if this is a critical feature for you, we can prioritize this or take a PR if you are willing to implement it!

**Comment by sunjayBhatia** [comment]
@artiommocrenco hey there, @tsaarni brought up an interesting point rereading the Ingress spec in [this comment](https://github.com/projectcontour/contour/pull/4595#issuecomment-1169730214)

```
The description for leaving field Ingress.spec.tls.secretName empty says "Field is left optional to allow TLS routing based on SNI hostname alone" so I wonder if that option is really meant for TLS passthrough?
```

However from your example it looks like you would rather expect to match the nginx behavior and have Contour/Envoy serve the fallback cert when an Ingress does not specify a TLS secret? Just wanted to clarify what ux people would expect here

**Comment by tsaarni** [comment]
I searched some historical discussion from Kubernetes github about empty `secretName` and found these:
* Discussion about the use case for leaving `secretName` empty https://github.com/kubernetes/kubernetes/pull/23500#issuecomment-201688415 (went into Kubernetes v1.2.0)
* Terminology used in printout: "routes" vs "**terminates**" https://github.com/kubernetes/kubernetes/blob/b1e130fe83156783153538b6d79821c2fdaa85bb/staging/src/k8s.io/kubectl/pkg/describe/describe.go#L2758-L2778

This seems bit under specified API, but in my opinion the little evidence there is points towards this being _meant to be used_ for TLS passthrough.

**Comment by youngnick** [comment]
One of the biggest problems with the Ingress spec has *always* been how under-specced it is, so I'm not surprised to find that this behavior is the same.

It seems like the issue for @artiommocrenco is that Contour doesn't work like ingress-nginx here, and that having it serve the fallback when the secret is missing is the request?

I think that, in the absence of a clear spec "Do what ingress-nginx does for Ingress" is probably the right play, since it is the defacto default Ingress controller.

tl;dr If we're going to support empty secret references in Ingress, I think it should supply the fallback certificate if one is configured, and not generate any configuration if one is not. That's the best we can do.

(Also note that this type of problem is why we are trying so hard to specify everything fully in Gateway API.)

**Comment by sunjayBhatia** [comment]
yeah I agree with the above

- when you say fallback is enabled via annotation + cert configuration in contour, and no tls cert you get the fallback served for terminating tls
- when you say fallback is enabled via annotation + cert configuration in contour, and tls configured on the ingress, tls is configured for terminating most requests you get the fallback served for requests w/ no SNI
- another additional thing to add: if no annotation and no cert on the ingress, right now my PR has it so the routes/vhost don't get configured, *we could configure passthrough TLS instead* or *add another annotation that enables passthrough, similar to HTTPProxy TLS options*

**Comment by sunjayBhatia** [comment]
another option to the implementation in https://github.com/projectcontour/contour/pull/4595 would be to have the value in the annotation `projectcontour.io/enable-fallback-certificate` be a comma separated list of hostnames that the fallback would be applied to, to give some more granular control over things, rather than enabling fallback support on all hosts configured the Ingress resource

**Comment by sunjayBhatia** [comment]
@artiommocrenco Hey there, have you had a chance to take a look at the discussion above and do you have any thoughts on how you expect things to work?

Contour's "fallback cert" was not intended to be a catchall/wildcard TLS cert the server presents in general to any request but rather specifically to cover the case of old TLS clients who misconfigure or don't include an SNI in the connection details. We can possibly stretch the meaning of this cert if need be (to basically match nginx ingress since the Ingress spec itself is a little under-specified), but would want to make sure the UX of what we provide (options described in a few comments above) makes sense.
